File: app/src/main/java/danb/speedrunbrowser/utils/LCSMatcher.java
package danb.speedrunbrowser.utils;

public class LCSMatcher {
    private String needle;
    private String haystack;
    private int threshold;

    private int[] haystackSubstringRemaining;

    private int maxSubstring;


    public LCSMatcher(String needle, String haystack, int threshold) {
        this.needle = needle;
        this.haystack = haystack;
        this.threshold = threshold;

        haystackSubstringRemaining = new int[haystack.length()];

        calculateLCS();
    }

    private void calculateLCS() {
        int[][] mat = new int[needle.length() + 1][haystack.length() + 1];

        for(int i = 1;i <= needle.length();i++) {
            for(int j = 1;j <= haystack.length();j++) {
                if(needle.charAt(i - 1) == haystack.charAt(j - 1)) {
                    mat[i][j] = mat[i - 1][j - 1] + 1;
                    haystackSubstringRemaining[j - mat[i][j]] = Math.max(haystackSubstringRemaining[j - mat[i][j]], mat[i][j]);

                    maxSubstring = Math.max(maxSubstring, haystackSubstringRemaining[j - mat[i][j]]);
                }
            }
        }
    }

    public int getSubstringRemaining(int pos) {
        return haystackSubstringRemaining[pos];
    }

    public int getMaxMatchLength() {
        return maxSubstring;
    }
}

File: app/src/main/java/danb/speedrunbrowser/utils/LCSMatcher.kt
package danb.speedrunbrowser.utils

class LCSMatcher(private val needle: String, private val haystack: String, private val threshold: Int) {

    private val haystackSubstringRemaining = IntArray(haystack.length)

    var maxMatchLength: Int = 0
        private set

    init {
        calculateLCS()
    }

    private fun calculateLCS() {
        val mat = Array(needle.length + 1) { IntArray(haystack.length + 1) }

        for (i in 1..needle.length) {
            for (j in 1..haystack.length) {
                if (needle[i - 1] == haystack[j - 1]) {
                    mat[i][j] = mat[i - 1][j - 1] + 1
                    haystackSubstringRemaining[j - mat[i][j]] = Math.max(haystackSubstringRemaining[j - mat[i][j]], mat[i][j])

                    maxMatchLength = Math.max(maxMatchLength, haystackSubstringRemaining[j - mat[i][j]])
                }
            }
        }
    }

    fun getSubstringRemaining(pos: Int): Int {
        return haystackSubstringRemaining[pos]
    }
}
